### PR TITLE
feat(frontend): evaluate expression

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
- add ExprImpl::eval_row_const, remove Literal::eval_as
- update scan range building logic. const instead of literal is supported. 
  - When the constant has a larger type and cannot be casted to the column's type, we can't convert it to scan range yet. e.g., `select * from t where v1 < 60000 -- v1 is smallint`. Ideally it can be converted but might be tricky, so not handled here. Previously this is actually a hidden bug (`eval_as` will panic), as described by @xiangjinwu [here](https://github.com/singularity-data/risingwave/issues/3479#issuecomment-1179801487)

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
https://github.com/singularity-data/risingwave/issues/3479 https://github.com/singularity-data/risingwave/issues/3477

### CODE REVIEW

File: src/frontend/test_runner/tests/testdata/range_scan.yaml [hunk 1]
```diff
@@ -27,16 +27,30 @@
...
-        BatchScan { table: orders_count_by_user, columns: [user_id, date, orders_count] }
+      BatchScan { table: orders_count_by_user, columns: [user_id, date, orders_count], scan_ranges: [user_id = Int64(43)] }
 - before:
     - create_table_and_mv
   sql: |
     SELECT * FROM orders_count_by_user WHERE user_id = 'a'
   batch_plan: |
     BatchExchange { order: [], dist: Single }
-      BatchFilter { predicate: ($0 = 'a':Varchar::Int64) }
+      BatchFilter { predicate: false:Boolean }
```
Comment (fuyufjh) [inline src/frontend/test_runner/tests/testdata/range_scan.yaml:37]: This can be further optimized out, but may be beyond the scope of this PR.

**Review by fuyufjh — APPROVED** [review]
LGTM

File: src/frontend/src/expr/mod.rs [hunk 2]
```diff
@@ -442,20 +458,21 @@ impl ExprImpl {
...
 
-    pub fn as_in_literal_list(&self) -> Option<(InputRef, Vec<Literal>)> {
+    pub fn as_in_const_list(&self) -> Option<(InputRef, Vec<ExprImpl>)> {
         if let ExprImpl::FunctionCall(function_call) = self &&
         function_call.get_expr_type() == ExprType::In {
             let mut inputs = function_call.inputs().iter().cloned();
             let input_ref= match inputs.next().unwrap() {
                 ExprImpl::InputRef(i) => *i,
-                _ => unreachable!()
+                _ => { return None }
```
Comment (xiangjinwu) [inline src/frontend/src/expr/mod.rs:467]: Let's add a test case for this bug fix. For example `user_id in (42.0, 44.0)`

File: src/frontend/src/utils/condition.rs [hunk 3]
```diff
@@ -290,24 +290,29 @@ impl Condition {
...
                     }
-                    eq_conds = vec![lit];
-                } else if let Some((input_ref, in_lit_list)) = expr.as_in_literal_list() {
+                    eq_conds = vec![value];
+                } else if let Some((input_ref, in_const_list)) = expr.as_in_const_list() {
                     assert_eq!(input_ref.index, order_column_ids[i]);
                     let mut scalars = vec![];
-                    for lit in in_lit_list.into_iter().unique() {
-                        if lit.is_null() {
+                    for const_expr in in_const_list.into_iter().unique() {
```
Comment (xiangjinwu) [inline src/frontend/src/utils/condition.rs:307]: `unique` here is less effective, as it compares the expression rather than the evaluated result.

**Review by xiangjinwu — COMMENTED** [review]
lgtm

**Pull request merged**